### PR TITLE
feat: allow cross-namespace locking for semaphore and mutex

### PR DESCRIPTION
**Pull request by woehrl01**
This Pull Request introduces the functionality to specify a namespace for the semaphore configmap references, allowing for locking on shared resources across namespaces.

### Motivation

Enhancing the semaphore configuration with a namespace property enables better resource management in environments with shared resources across multiple namespaces.

### Modifications

Introduced a `namespace` property to the semaphore configuration. This property is designed to specify the namespace in which the semaphore configmap resides.

#### Design choices:
* The namespace is not directly added to the `ConfigMapKeyRef` in an attempt to avoid violating the official definition and to prevent potential misunderstandings. This design decision was made to keep the definition of `ConfigMapKeyRef` as standard and clear as possible.

### Verification

Added a unit test

### CODE REVIEW

**Comment by woehrl01** [comment]
@terrytangyuan @jessesuen @alexec highly appreciate your feedback on this. Is there anything missing, to have this merged? Thanks!

**Comment by terrytangyuan** [comment]
cc @sarabala1979 @isubasinghe

**Review by isubasinghe — COMMENTED** [review]
Looks good to me. I think this PR should also implement this support for mutexes though. 
It would be somewhat odd if semaphores supported this while mutexes didn't.

**Comment by isubasinghe** [comment]
The tests pass for me locally, might just be flaky tests, give it a re-run

**Comment by woehrl01** [comment]
Thanks for your feedback @isubasinghe I just added the namespace for mutex, too.

File: api/jsonschema/schema.json [hunk 1]
```diff
@@ -5416,6 +5416,10 @@
         "name": {
           "description": "name of the mutex",
           "type": "string"
+        },
+        "namespace": {
+          "description": "Namespace is the namespace of the configmap if not the same as the workflow",
+          "type": "string"
```
Comment (isubasinghe) [inline api/jsonschema/schema.json:5422]: Namespace is the namespace of the mutex if not the same as the workflow.

File: api/openapi-spec/swagger.json [hunk 2]
```diff
@@ -9352,6 +9352,10 @@
         "name": {
           "description": "name of the mutex",
           "type": "string"
+        },
+        "namespace": {
+          "description": "Namespace is the namespace of the configmap if not the same as the workflow",
```
Comment (isubasinghe) [inline api/openapi-spec/swagger.json:9357]: .

File: docs/executor_swagger.md [hunk 3]
```diff
@@ -2468,6 +2468,7 @@ than the MaxAge, it will be ignored. |  |
 | Name | Type | Go type | Required | Default | Description | Example |
 |------|------|---------|:--------:| ------- |-------------|---------|
 | name | string| `string` |  | | name of the mutex |  |
+| namespace | string| `string` |  | | Namespace is the namespace of the configmap if not the same as the workflow |  |
```
Comment (isubasinghe) [inline docs/executor_swagger.md:2471]: .

File: docs/fields.md [hunk 4]
```diff
@@ -2338,6 +2338,7 @@ Mutex holds Mutex configuration
 | Field Name | Field Type | Description   |
 |:----------:|:----------:|---------------|
 |`name`|`string`|name of the mutex|
+|`namespace`|`string`|Namespace is the namespace of the configmap if not the same as the workflow|
```
Comment (isubasinghe) [inline docs/fields.md:2341]: .

File: pkg/plugins/executor/swagger.yml [hunk 5]
```diff
@@ -2143,6 +2143,10 @@ definitions:
       name:
         description: name of the mutex
         type: string
+      namespace:
+        description: Namespace is the namespace of the configmap if not the same as
```
Comment (isubasinghe) [inline pkg/plugins/executor/swagger.yml:2147]: .

**Comment by woehrl01** [comment]
Yes, you're right, I rephrase the comment. Thank you!

File: docs/executor_swagger.md [hunk 6]
```diff
@@ -2468,6 +2468,7 @@ than the MaxAge, it will be ignored. |  |
 | Name | Type | Go type | Required | Default | Description | Example |
 |------|------|---------|:--------:| ------- |-------------|---------|
 | name | string| `string` |  | | name of the mutex |  |
+| namespace | string| `string` |  | `"namespace of workflow)"`|  |  |
```
Comment (isubasinghe) [inline docs/executor_swagger.md:2471]: fix default here

File: docs/executor_swagger.md [hunk 7]
```diff
@@ -4229,6 +4230,7 @@ Note that this field cannot be set when spec.os.name is windows.
 | Name | Type | Go type | Required | Default | Description | Example |
 |------|------|---------|:--------:| ------- |-------------|---------|
 | configMapKeyRef | [ConfigMapKeySelector](#config-map-key-selector)| `ConfigMapKeySelector` |  | |  |  |
+| namespace | string| `string` |  | `"namespace of workflow)"`|  |  |
```
Comment (isubasinghe) [inline docs/executor_swagger.md:4233]: also here, there is a right paren and its inside `` and "".

**Review by isubasinghe — COMMENTED** [review]
typo in docs, I made two comments

**Review by isubasinghe — CHANGES_REQUESTED** [review]
Apoligies: meant to request changes not comment, see previous comment.

**Comment by woehrl01** [comment]
Just renamed the default again and put it in brackets to be more visible of a "variable" argument. Unfortunately, I can't remove the quotes, they are added automatically. Alternative would be to put the default into the description, but I'm not a fan of that either.

**Comment by isubasinghe** [comment]
Thanks for the changes @woehrl01

**Review by isubasinghe — APPROVED** [review]
LGTM

**Pull request merged**